File: Gdx/fell/core/src/com/dugsolutions/fell/map/MapSquareUI.java
package com.dugsolutions.fell.map;

import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Mesh;
import com.badlogic.gdx.graphics.VertexAttribute;
import com.badlogic.gdx.graphics.VertexAttributes;
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion;
import com.badlogic.gdx.graphics.glutils.ShaderProgram;
import com.badlogic.gdx.math.Matrix4;

public class MapSquareUI extends MapSquareData {
	// This square is displaying this region in the atlas.
	AtlasRegion region;
	Mesh mesh;

	public MapSquareUI(AtlasRegion r) {
		region = r;
	}

	public void build(float startX, float startY, float width, float height,
			float startZ, float scaleZ) {
		build(startX, startY, width, height, region.getU(), region.getV(),
				region.getU2(), region.getV2(), startZ, scaleZ);

		mesh = new Mesh(true, verts.length / 6, indices.length,
				new VertexAttribute(VertexAttributes.Usage.Position, 3,
						ShaderProgram.POSITION_ATTRIBUTE), new VertexAttribute(
						VertexAttributes.Usage.ColorPacked, 4,
						ShaderProgram.COLOR_ATTRIBUTE), new VertexAttribute(
						VertexAttributes.Usage.TextureCoordinates, 2,
						ShaderProgram.TEXCOORD_ATTRIBUTE + "0"));

		mesh.setVertices(verts);
		mesh.setIndices(indices);
	}

	public void render(ShaderProgram program, Matrix4 projMatrix) {
		region.getTexture().bind();
		program.begin();
		program.setUniformMatrix("u_projTrans", projMatrix);
		program.setUniformi("u_texture", 0);
		mesh.render(program, GL20.GL_TRIANGLES);
		program.end();
	}
}

File: Gdx/FellDungeon/app/src/main/java/com/dugsolutions/fell/map/MapSquareUI.kt
package com.dugsolutions.fell.map

import com.badlogic.gdx.graphics.*
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion
import com.dugsolutions.fell.map.MapSquareData
import com.badlogic.gdx.graphics.glutils.ShaderProgram
import com.badlogic.gdx.math.Matrix4

class MapSquareUI(  // This square is displaying this region in the atlas.
    var region: AtlasRegion
) : MapSquareData() {

    var mesh: Mesh? = null

    fun build(
        startX: Float, startY: Float, width: Float, height: Float,
        startZ: Float
    ) {
        build(
            startX, startY, width, height, region.u, region.v,
            region.u2, region.v2, startZ
        )
        val mesh = Mesh(
            true, verts.size / 6, indices.size,
            VertexAttribute(
                VertexAttributes.Usage.Position, 3,
                ShaderProgram.POSITION_ATTRIBUTE
            ), VertexAttribute(
                VertexAttributes.Usage.ColorPacked, 4,
                ShaderProgram.COLOR_ATTRIBUTE
            ), VertexAttribute(
                VertexAttributes.Usage.TextureCoordinates, 2,
                ShaderProgram.TEXCOORD_ATTRIBUTE + "0"
            )
        )
        this.mesh = mesh
        mesh.setVertices(verts)
        mesh.setIndices(indices)
    }

    fun render(program: ShaderProgram, projMatrix: Matrix4?) {
        region.texture.bind()
        program.begin()
        program.setUniformMatrix("u_projTrans", projMatrix)
        program.setUniformi("u_texture", 0)
        mesh?.render(program, GL20.GL_TRIANGLES)
        program.end()
    }

    override fun setElevations(startX: Float, startY: Float, width: Float, height: Float, startZ: Float) {
        super.setElevations(startX, startY, width, height, startZ)
        mesh?.setVertices(verts)
    }
}